Expanded Utility functions

gitMyPad opened the pull request:
Fixed native timers (preserving correct behavior in cases where the natives fail.)
Added UnitGroup to handle folder. (For a more efficient ForGroup)
Added World and UnitEvent to utils folder.
Updated Catalog to reflect the changes made by these additions.

gitMyPad committed:
Update Timer.lua

Overridden a lot of the Timer natives.

gitMyPad committed:
Add files via upload

Uploaded UnitEvent and World

gitMyPad committed:
Update Catalog.lua

Added UnitGroup, World, and UnitEvent
Removed Execute, Init, Log, Native, Optimizations, and Rawcode from Catalog.

gitMyPad committed:
Add files via upload

Added UnitGroup (with an overridden ForGroup)

Trokkin committed:
Merge branch 'master' into master

Trokkin commented on handle/UnitGroup.lua:
```diff
@@ -0,0 +1,21 @@
+
+    while (i < j) do
+        __u = BlzGroupUnitAt(g, i)
+        pcall(func)
```
pcall _returns_ error if it happens. Use `execute` from base.Execute to handle errors, so that catched errors are displayed.
But I don't recall `ForGroup` managing errors, so I suggest removing `pcall` from here.

gitMyPad replied:
Will do. (Calling function directly instead of execute)

Trokkin commented on handle/UnitGroup.lua:
```diff
@@ -0,0 +1,21 @@
+    return __u
+end
+
+replaceNative('GetEnumUnit', tempEnumUnit)
```
=> `replaceNative('getenumunit', function() return __u end)`?

gitMyPad replied:
Okay. Will inline that.

Trokkin commented on handle/UnitGroup.lua:
```diff
@@ -0,0 +1,21 @@
+require('CeresStdLib.base.Native')
+
+local __u
```
I'm not sure if using one local is the best solution now, since we're in Lua.
Exactly, you can modify local environment of the called function, where you then define `GetEnumUnit()` referencing your current unit.

Trokkin commented on handle/Timer.lua:
```diff
@@ -1,30 +1,190 @@
-local free_timers = {}
+
+local Timer 		= {}
+local free_timers	= {}
```
I never heard that native timer functions fail. Could you evaluate on such cases a little, maybe provide failing test case?

gitMyPad replied:
Issues with Repeating Timers failing to repeat once they are resumed (after a pause).
There's also the ResumeTimer thing which can crash the game if called before TimerStart was invoked for the timer object.
There was a thread in the Lab section of Hive delving into these things.
https://www.hiveworkshop.com/threads/issues-with-timer-functions.309433/

The Timer.lua was overwritten to address at least those two above.

Trokkin commented on util/UnitEvent.lua:
```diff
@@ -0,0 +1,186 @@
+    Uses Magic Defense instead of Defend ability to detect leave.    
+]]
+
+require('CeresStdLib.base.Rawcode')
```
Use base.Basics instead of all those.

gitMyPad replied:
Ok, will fix.

Trokkin commented on util/UnitEvent.lua:
```diff
@@ -0,0 +1,186 @@
+
+local trigTable         = {}
+local __userData        = {}
+__userData[-1]          = {}   --  Holds references to each unit
```
Why not just another local var? At least, why not unique name inside `__userdata`?

Trokkin commented on util/World.lua:
```diff
@@ -0,0 +1,21 @@
+World               = {}
+local metaWorld     = {
+    __newindex      = function(t, k, v)
+        k = nil
```
k, v are local variables and do not require such operations. `function() end` suits this better.

gitMyPad replied:
Just wanted to make sure that no write operation is actually being done. Will simplify.

Trokkin commented on util/World.lua:
```diff
@@ -0,0 +1,21 @@
+    World.MAX_Y     = GetRectMaxY(World.RECT)
+    World.MIN_X     = GetRectMinX(World.RECT)
+    World.MIN_Y     = GetRectMinY(World.RECT)
+    setmetatable(World, metaWorld)
```
Also, if this metatable suits only the purpose of not modifying the object, I suggest moving it as a unique module into .base scope, so it can be reused for other const objects.
Could be named `base.Const` or something like that.

Trokkin commented on util/UnitEvent.lua:
```diff
@@ -0,0 +1,186 @@
+local __initialized     = false
+
+replaceNative('GetUnitUserData', function(u)
+    return __userData[u]
```
I heard handles in lua fail to be consistent keys in tables. Consider referencing them by id.

gitMyPad replied:
So, by using GetHandleId? Will try to write that up.

Trokkin replied:
maybe, but maybe better using `GetUnitID`.

Trokkin commented on util/UnitEvent.lua:
```diff
@@ -0,0 +1,186 @@
+UnitEvent = {}
+
+local trigTable         = {}
+local __userData        = {}
```
No need to prefix `__` each local variable, by the way.

gitMyPad replied:
Sure

Trokkin reviewed (changes requested):
Some moments require changes, but overall good job! Looking forward to merge our first PR!
_side note_: 2 out of 4 module changes are small and pretty, could be merged almost instantly if they had a unique PR for each.

gitMyPad committed:
Update UnitGroup.lua

Removed pcall to func in overridden native ForGroup

Trokkin commented on handle/UnitGroup.lua:
```diff
@@ -1,21 +1,17 @@
 require('CeresStdLib.base.Native')
 
-local __u
+local __u = u
```
This seems like an error or something, there is no such global named `u`

gitMyPad replied:
Oops, that one was accidental.

Trokkin commented on util/UnitEvent.lua:
```diff
@@ -0,0 +1,186 @@
+    __userData[i], __userData[0]         = __userData[0], i
+end
+
+function GetUnitId(u) return Native.GetUnitUserData(u) end
```
Do we have a reason to have recyclable IDs now? They were implemented as a necessary evil to be able to fit unit-unique data into strictly sized JASS arrays, but now we only have maps, so HandleId do the job probably even better, as they're managed by the game itself.

gitMyPad replied:
They were included out of habit when I first wrote it up. Would the best course of action be to remove them?

Trokkin commented on handle/Timer.lua:
```diff
@@ -1,30 +1,190 @@
+local free_timers	= {}
+local MAX_PUSHES	= 10000		--	Maximum amount of times push is called before garbage collection.
+
+free_timers[0]		= 0
```
Again, use unique names.

gitMyPad replied:
replaced [0] with .free, and [1] with .steps

Trokkin commented on handle/Timer.lua:
```diff
@@ -1,30 +1,190 @@
+free_timers[-1]		= 0
+
+free_timers.has = function()
+	if free_timers[0] <= 0 then return false end return true
```
== `return free_timers[0] > 0` lol

gitMyPad replied:
inlined

Trokkin commented on handle/Timer.lua:
```diff
@@ -1,30 +1,190 @@
+		looped      = nil,
+		running     = nil,
+		inCallback  = nil,
+		pauseFlag   = 0,
```
Flags should be boolean?

gitMyPad replied:
Indicated now.

Trokkin commented on handle/Timer.lua:
```diff
@@ -1,30 +1,190 @@
+
+replaceNative('PauseTimer', function(t)
+	if Timer[t].running then
+		Timer[t].running    = nil
```
Same here, timer is a handle thus keys should be referencing it by ID.

gitMyPad replied:
Replaced all keys being timers with their handle ids.

Trokkin commented on handle/Timer.lua:
```diff
@@ -1,30 +1,190 @@
+replaceNative('TimerStart', function(t, dur, looper, func)
+	if not Timer[t].hasCallback then
+		--  Make it come true
+		Timer[t].hasCallback = 0
```
Then what's the point of the check at all? Remove both if closures and leave only this assignment

gitMyPad replied:
Removed the first if condition.

Trokkin commented on handle/Timer.lua:
```diff
@@ -1,30 +1,190 @@
+
+function TimerSetRemaining(t, newR, update)
+	if Timer[t].inCallback then
+		Log.warn('TimerSetRemaining: The remaining duration of the timer ' .. I2S(GetHandleId(t)) .. ' cannot be altered while the callback function is running.')
```
Is this not an error?

gitMyPad replied:
I thought a warning would be more appropriate here, given that the user is responsible for calling this function. For errors, I think that they should be on the system itself, and not the user.

Trokkin commented on handle/Timer.lua:
```diff
@@ -1,30 +1,190 @@
+	else
+		obj = Native.CreateTimer()
+	end
+	Timer[obj]      = {
```
Are all of these fields necessary? Overall it feels a bit overcomplicated.
I still can merge it, and try myself at simplifying it, though.

gitMyPad replied:
The fields describe the current state of the timer.
hasCallback indicates when a callback is added to the timer (via TimerStart). This is for the ResumeTimer crash fix.
inCallback indicates when the timer has expired and is now calling the callback.
callback is the callback function of the timer.

Trokkin commented on handle/Timer.lua:
```diff
@@ -1,30 +1,190 @@
+replaceNative('TimerGetTimeout', function(t) return Timer[t].duration end)
+replaceNative('TimerGetElapsed', function(t) return Native.TimerGetElapsed(t) + Timer[t].elapsed end)
+
+function TimerSetRemaining(t, newR, update)
```
Do we need this function at all? Considering the now complexity of your native replacements, it could be worth itself, but I wonder if you could simplify everything so this won't be needed.

gitMyPad committed:
Update UnitEvent.lua

No longer uses units as keys. Instead, their handle ids are used.
Removed "__" prefix.
No longer overrides the natives GetUnitUserData and SetUnitUserData

gitMyPad committed:
Update UnitEvent.lua

Added more description on top.

gitMyPad committed:
Updated Timer.lua

Replaced instances of [0] to .free in table free_timers
Replaced instances of [-1] to .steps in table free_timers
Now stores the attributes of the timer within the handle id.

gitMyPad committed:
Update Timer.lua

Inlined free_timers.has

gitMyPad committed:
Update UnitGroup.lua

gitMyPad committed:
Update Basics.lua

gitMyPad committed:
Update Basics.lua

Added a new (optional) parameter in function makeReadonly (for debug purposes).

gitMyPad committed:
Update World.lua

Made the script require Basics.
Now uses the makeReadonly function to make World a readonly table object.

gitMyPad committed:
Moved META_TABLE to Const

gitMyPad committed:
Updates

Updated UnitGroup to be a readonly table.
Updated UnitEvent to use GetUnitId internally

gitMyPad committed:
Merge remote-tracking branch 'upstream/master'

gitMyPad committed:
Added new file Objects

Updated Timer.lua to avoid ResumeTimer bug when called within callback.

Trokkin commented on base/Basics.lua:
```diff
@@ -4,15 +4,22 @@ require('CeresStdLib.base.Rawcode')
 DUMMY_PLAYER = players[PLAYER_NEUTRAL_PASSIVE]
 DUMMY_HOSTILE_PLAYER = players[PLAYER_NEUTRAL_AGGRESSIVE]
+
+local __oldInitB = InitBlizzard
```
1. Can it ever fail?
2. Use replaceNative

gitMyPad replied:
For some reason, in my copy of the game, that function appeared to be faulty, causing the thread to terminate before it can reach InitGlobals and so on.

Trokkin commented on Catalog.lua:
```diff
@@ -1,14 +1,8 @@
--- require('CeresStdLib.Catalog')
-
 require('CeresStdLib.base.Basics')
-require('CeresStdLib.base.Native')
```
The reason for Catalog to have *all* the modules is for user to easily pick them to c&p, so please return them back here.

gitMyPad replied:
Looks like I overlooked that one.

Trokkin commented on handle/Objects.lua:
```diff
@@ -0,0 +1,2 @@
+require('CeresStdLib.handle.Timer')
```
This entire file is unnecessary, at least while there's nothing general for every handle.

Trokkin commented on base/Const.lua:
```diff
@@ -0,0 +1,27 @@
+local META_TABLE    = {
```
I suggest `makeReadonly` to create new local metatable with functions unique to each call, referencing local variables from the call. Having a global reference table may be necessary in some places, but we don't need to do it literally every time now.

Trokkin commented on handle/Timer.lua:
```diff
@@ -1,37 +1,210 @@
+replaceNative('ResumeTimer', function(t)
+	local i = GetHandleId(t)
+	if not Timer[i].hasCallback then
+		Log.error("ResumeTimer: Attempted to resume a timer " .. I2S(i) .. " with no callback")
```
Lua does number to string conversion implicitly, so no need for `I2S()`.

Trokkin commented on handle/Timer.lua:
```diff
@@ -1,37 +1,210 @@
+	local i = GetHandleId(t)
+	if Timer[i].running then
+		Timer[i].running    = nil
+		Timer[i].pauseFlag  = (BlzBitOr(Timer[i].pauseFlag, FLAG_TIMER_PAUSED))
```
It seems that WC run Lua 5.3, which [has bitwise operation implemented natively](https://www.lua.org/manual/5.3/manual.html#3.4.2). Check and replace `BlzBitOr()` if possible.

gitMyPad replied:
Didn't want to use them at first, so kept those bit operations. Replacing them later.

Trokkin commented on handle/Timer.lua:
```diff
@@ -1,37 +1,210 @@
 function doPeriodicaly(period, func)
-	local t = CreateTimer()
-	TimerStart(t, period, true, function() func(t) DestroyTimer(GetExpiredTimer()) end)
+	TimerStart(CreateTimer(), period, true, function() func(GetExpiredTimer()) DestroyTimer(GetExpiredTimer()) end)
```
I think the last merge was incorrect, `local t` shouldn't be inlined back.
Also, this `DestroyTimer()` shouldn't be here (my own stupidity), so please remove it.

gitMyPad replied:
Will remove.

Trokkin commented on util/UnitEvent.lua:
```diff
@@ -0,0 +1,165 @@
+userData.nativeHasAbil  = {}   --  Holds a flag telling the system if the unit already has that ability.
+userData.indexed        = {}
+
+trigTable[0]            = CreateTrigger()
```
again, use readable names. `[0]` and such tells nothing.

Trokkin replied:
also, you don't need actual table for it.

Trokkin commented on util/UnitEvent.lua:
```diff
@@ -0,0 +1,165 @@
+require('CeresStdLib.handle.UnitGroup')
+require('CeresStdLib.util.World')
+
+EVENT_UNIT_ENTER        = 1
```
Should be a unique object. `= {}` does the trick.

Trokkin commented on util/UnitEvent.lua:
```diff
@@ -0,0 +1,165 @@
+end)
+
+ceres.addHook('main::after', function()
+    GroupEnumUnitsInRect(trigTable[2], World.RECT, nil)
```
If I recall correctly, there's no way to have units created before the game starts without `CreateUnit` native, so I suggest replacing the native and not doing the loop.

gitMyPad replied:
Considering that, I'd have to replace CreateCorpse as well.

gitMyPad committed:
Refresh button

Pucci must reset the universe

gitMyPad committed:
Made files lowercase

Made all files in the Stdlib lowercase.

gitMyPad committed:
Rename Rawcode.lua to rawcode.lua

Trokkin commented:
13 files are kind of a headache to review and approve at once
Please, if you push something different than polishing already reviewed code, make a separate PR for it

gitMyPad committed:
Updated Timer.lua

Rewritten Timer to wrap native timers in a "class".

gitMyPad committed:
Merge branch 'master' of https://github.com/MediumRareSteak-Hive/CeresStdLib

gitMyPad committed:
Rename Timer.lua to timer.lua

gitMyPad committed:
Rename UnitGroup.lua to unitgroup.lua

gitMyPad committed:
Rename Benchmark.lua to benchmark.lua

gitMyPad committed:
Rename World.lua to world.lua

gitMyPad committed:
Rename UnitEvent.lua to unitevent.lua

gitMyPad committed:
Rename Shell.lua to shell.lua

gitMyPad committed:
Rename Optimizations.lua to optimizations.lua

gitMyPad committed:
Rename Native.lua to native.lua

gitMyPad committed:
Rename Log.lua to log.lua

gitMyPad committed:
Rename Execute.lua to execute.lua

gitMyPad committed:
Rename Basics.lua to basics.lua

gitMyPad committed:
Rename Init.lua to init.lua

gitMyPad committed:
Added Handle.lua (Experimental)

gitMyPad committed:
Merge branch 'master' of https://github.com/MediumRareSteak-Hive/CeresStdLib

gitMyPad committed:
Update timer.lua

Fixed Timer.__index and Timer.__newindex

Trokkin commented:
What will happen to this PR when handle shadowing will be merged or vice versa?
Maybe this PR isn't relevant anymore?

gitMyPad commented:
I suppose this will have to be closed. I made too many changes to the handles folder and to an extent, the utils folder for it to be merged here (most of my commits are in the fixes branch.)

Trokkin commented:
Done.
Your effort buried in here will not be forgotten.
